synapse/util/task_scheduler.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import TYPE_CHECKING, Awaitable, Callable, Dict, List, Optional, Set, Tuple
+
+from twisted.python.failure import Failure
+
+from synapse.metrics.background_process_metrics import run_as_background_process
+from synapse.types import JsonMapping, ScheduledTask, TaskStatus
+from synapse.util.stringutils import random_string
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+logger = logging.getLogger(__name__)
+
+
+class TaskScheduler:
+    # Precision of the scheduler, evaluation of tasks to run will only happen
+    # every `SCHEDULE_INTERVAL_MS` ms
+    SCHEDULE_INTERVAL_MS = 5 * 60 * 1000  # 5mn
+    # Time before a complete or failed task is deleted from the DB
+    KEEP_TASKS_FOR_MS = 7 * 24 * 60 * 60 * 1000  # 1 week
```

Comment:
Why wait an arbitrary week to delete the task?

Reply:
I can add a field, but I think my arbitrary value should cover 90% of the use cases, and such a field could be added later if the need arises ?

Reply:
Some task based architecture I've used will store the result until it is fetched and then delete it immediately. (You do still need a clean-up task if some results are never used.)

Reply:
I am not sure exactly what you are suggesting. Just add a method to delete a task, and keep the current clean-up part?

Reply:
I'm wondering if our "get result" should be "get and delete result" -- it might not matter much, but it depends on how often we expect things to go into this table.

Reply:
Ohh ok. I don't think so, because:
- there is no real get_result, we just fetch the task descriptor that includes the results and status
- in the purge rooms use case, I am taking advantage of that to NOT store the purges in another table

I am going to add an explicit `delete_task` however.